host/docs/manpages/uhd_adc_self_cal.1
.TH UHD_ADC_SELF_CAL "1" "July 2021" "uhd_adc_self_cal 4.1" "User Commands"
.SH NAME
uhd_adc_self_cal \- invoke calibration in some UHD devices.
.SH DESCRIPTION
.SS "UHD ADC self calibration Allowed options:"
.TP
\fB\-\-help\fR
help message
.TP
\fB\-\-version\fR
print the version string and exit
.TP
\fB\-\-args\fR arg
device address args, using usual UHD device selection specifiers.


CHANGELOG
Change Log for Releases
==============================

## 004.009.000.001

Changes compared to UHD 4.9.0.0:

* Dependency changes
  - Support Boost 1.89.0
  - Drop support for boost below 1.71.0
* Fixes
  - Fix static build of libuhd
* Notes
  - We suggest to only update to this version if the boost 1.89.0 compability
    is required. Apart from the Boost build fixes, there are no other changes
    in this UHD release.

## 004.009.000.000
* Highlights / Main Changes
  - Support for USRP B206mini-i
  - Support for OBX daughterboard
* New Features
  - OBX daughterboard and B206mini support added.
  - X3x0: Add support for the gps_iface feature. This allows sending arbitrary
    commands to the GPS module.
  - rfnoc_modtool: Add a new subcommand, `make-yaml`, that generates a block
    YAML file from some user inputs. When calling `rfnoc_modtool add`  without
    a valid block YAML file, the tool will now automatically call `make-yaml` to
    generate a new YAML file.
  - Added support for processing multiple samples per cycle in FFT block.
  - LabVIEW/USRP: New repository containing example applications which 
    demonstrate how to use NI LabVIEW with USRP devices via the UHD APIs.
    (https://github.com/ni/labview-usrp-examples)
  - Improved RFNoC OOT Block support:
    New repository containing official RFNoC OOT Blocks
    (https://github.com/EttusResearch/rfnoc-oot-blocks)
    - Added new RFNoC Aurora Block for TX and RX remote streaming applications
      to and from Aurora capable compute nodes like FPGA Coprocessors.
  - Enhanced the UHD Windows installer to automatically install Windows
    driver packages into the OS, eliminating this manual post-installation
    step for B-Series USB devices. 
* API Changes
  - `uhd::math::gcd` and `uhd::math::lcm` are removed from the public
     headers. Use `std::gcd` and `std::lcm` instead if C++17 is available,
     or `boost::math::gcd` and `boost::math::lcm` if not.
  - When calling `recv()` on an RX streamer connected to an RFNoC device, we
    no longer artificially inflate the timeout value. The old behaviour was
    in some cases abused to detect overruns, but this is not a reliable method
    and is now explicitly declared unsupported.
* Bug Fixes
  - RFNoC
    - Calling `recv()` on an RX streamer connected to an RFNoC device with a
      zero timeout value now correctly returns immediately.
    - rfnoc-gain: Moved the fpga/ subdirectory under rfnoc/. This way, the
      directory structure inside the OOT module is identical to the directory
      structure after installation.
  - X410:
    - Fixed NCO reset mechanism that caused ADC self-cal errors to occur during
      reopening of a UHD session when timed tune commands were previously used.
* Removals / Deprecations
  - Removed tools/uhd_dump. This tool is obsolete, the USRP X3x0 is not even
    compatible with it any more. Dumps can be recorded and analyzed with wireshark
    and the UHD dissectors in here.
  - Removed network_relay example.
* Validated OS Environments
  (Versions for build and runtime dependencies can be determined from the
   docker container definitions in the UHD repository at .ci/docker/...)
  - Linux:
    - Ubuntu: focal (20.04), jammy (22.04), noble (24.04), plucky (25.04)
    - Fedora: 40, 41, 42
  - Windows: 10 22H2, 11 23H2
  - MacOS: Sequoia (15.2)

## 004.008.000.000
* Highlights / Main Changes
  - Addition of rfnoc_modtool and further improvements to RFNoC Image Builder
    toolsets. This further improves the ability to design RFNoC blocks and
    create custom FPGA bitfiles.
  - Replaced RFNoC FFT block with a new version
  - Improved Windows installation capabilities including distribution
    of binary Python bindings for Windows though PyPI
* New Features
  - Image Builder
    - Add GRC support. This allows designing RFNoC bitfiles from GNU Radio
      Companion (GRC).
    - Improve clock connection checks and checks for duplicate connections.
    - Improve IO port compat check (e.g., check if wire widths match).
    - Allow default clocks in domain checks.
    - Simplify image core YAMLs by better usage of RADIO_NIPC parameter.
    - Improve parameter resolution.
    - Improve error messages for missing YAML files.
    - Add --SYNTH and --CHECK options.
    - Add support for building an FPGA image using multiple parallel jobs and
      unique seeds with repeat_fpga_build.py and the --fpga-jobs option.
    - Add 'q' (or 'quote') filter to YAML files to allow quoting strings.
  - rfnoc-gain (OOT RFNoC example)
    - Overhaul directory structure, and rename to rfnoc-gain from
      rfnoc-example.
    - Simplify dynamic loading of OOT DLLs by using the new modules.d feature.
    - Add CE clock support to gain block. This enables the example on X4x0.
    - Add a GNU Radio subdirectory with examples of how to run the gain block
      in GNU Radio.
  - Add rfnoc_modtool. This is a command line utility to help design OOT RFNoC
    blocks and replaces the previously available blocktool.
  - RFNoC
    - Add tune requests. This allows tuning a complete graph as known from
      multi_usrp instead of single blocks individually.
    - Replace FFT block with a new version supporting larger FFT sizes and
      cyclic prefix insertion and removal.
  - multi_usrp
    - Add Python bindings for get_user_settings_iface()
  - General UHD
    - Add modules.d support
    - Add uhd::find_uhd_command() API call
    - Remove old-style Boost.Asio usage for compatibility with future versions
      of Boost.
  - Embedded Filesystem
    - Support for USB Audio devices on E31X, E320, N3XX and X4XX
  - Windows binary installer
    - Simplified the Windows installer experience by reducing the number of
      individual downloads from 4 (for USB devices) to 2, by including
      libusb.dll and USB driver installer files into NSIS-built installer
      binaries.
    - Add new cmake feature option called ENABLE_EXTEND_WIN_PACKAGING,

host/utils/uhd_adc_self_cal.cpp
//
// Copyright 2020 Ettus Research, a National Instruments Brand
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <uhd/features/adc_self_calibration_iface.hpp>
#include <uhd/rfnoc/radio_control.hpp>
#include <uhd/rfnoc_graph.hpp>
#include <uhd/utils/safe_main.hpp>
#include <uhd/version.hpp>
#include <boost/program_options.hpp>
#include <iostream>
#include <sstream>

namespace po = boost::program_options;
using namespace uhd;

/****************************************************************************
 * main
 ***************************************************************************/
int UHD_SAFE_MAIN(int argc, char* argv[])
{
    po::options_description desc("Allowed options");
    // clang-format off
    desc.add_options()
        ("help", "help message")
        ("version", "print the version string and exit")
        ("args", po::value<std::string>()->default_value(""), "device address args")
    ;
    // clang-format on

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // print the help message
    if (vm.count("help")) {
        std::cout << "UHD ADC self calibration " << desc << std::endl;
        return EXIT_FAILURE;
    }

    if (vm.count("version")) {
        std::cout << uhd::get_version_string() << std::endl;
        return EXIT_SUCCESS;
    }
    const uhd::device_addr_t args  = vm["args"].as<std::string>();
    rfnoc::rfnoc_graph::sptr graph = rfnoc::rfnoc_graph::make(args);

    size_t num_calibrations = 0;
    for (auto radio_id : graph->find_blocks("Radio")) {
        auto radio_blk = graph->get_block<uhd::rfnoc::radio_control>(radio_id);
        if (radio_blk->has_feature<uhd::features::adc_self_calibration_iface>()) {
            auto& feature =
                radio_blk->get_feature<uhd::features::adc_self_calibration_iface>();

            const std::string ch_list = args.get("cal_ch_list", "");

            // Run it on all (enabled) channels
            const size_t num_channels = radio_blk->get_num_output_ports();
            for (size_t i = 0; i < num_channels; i++) {
                auto abs_ch =
                    radio_blk->get_block_id().get_block_count() * num_channels + i;
                if (ch_list.size() == 0
                    or ch_list.find(std::to_string(abs_ch)) != std::string::npos) {
                    std::cout << "Calibrating on channel " << i << " of " << radio_id
                              << "..." << std::endl;
                    feature.run(i, vm["args"].as<std::string>());
                    std::cout << "Finished!" << std::endl;
                    num_calibrations++;
                }
            }
        }
    }
    if (num_calibrations > 0) {
        std::cout << "Calibrated " << num_calibrations << " channels" << std::endl;
    } else {
        std::cerr << "WARNING: Did not find any channels to calibrate!" << std::endl;
    }

    return EXIT_SUCCESS;
}
